add blog post 2025-02-14 on cloud controller managers and the chicken and egg problem (#49612)
kubernetes/website: content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

Reviewer: Links within the site should be site-relative, eg `/blog/2024/05/20/completing-cloud-provider-migration/`.

Author: does this work for links in `kubernetes.io/docs` and `kubernetes.io/images` ?

Reviewer: This is kind of just how URIs work on the web.

To link to https://kubernetes.io/docs/home/, write `/docs/home/`

To link to static content, omit the `https://kubernetes.io` prefix (same thing)

Reviewer: We don't currently do this but if we need to write render-time hook code, we'll make that code handle cases correctly.

2. **API server network proxy** ([KEP-1281](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1281-network-proxy))  
3. **kubelet credential provider plugins** ([KEP-2133](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2133-kubelet-credential-providers))  
4. **Storage migration to use [CSI](https://github.com/container-storage-interface/spec?tab=readme-ov-file#container-storage-interface-csi-specification-)** ([KEP-625](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/625-csi-migration/README.md))

The [cloud controller manager is part of the control plane](https://kubernetes.io/docs/concepts/architecture/cloud-controller/). It is a critical component that replaces some functionality that existed previously in the kube-controller-manager and the kubelet.

![Components of Kubernetes](https://kubernetes.io/images/docs/components-of-kubernetes.svg)

One of the most critical functionalities of the cloud controller manager is the node controller, which is responsible for the initialization of the nodes.

> As we can see in the following diagram, when the **kubelet** starts, it registers the `Node` object with the apiserver, Tainting the node so it can be processed first by the cloud-controller-manager. The initial `Node` is missing the cloud-provider specific information, like the Node Addresses and the Labels with the cloud provider specific information like the Node, Region and Instance type information.

Reviewer: We avoid “we” unless you specifically mean “we,   Antonio Ojea & Michael McCune”.

> ![Diagram of Node initialization](https://mermaid.ink/img/pako:eNqVU01PAjEQ_StNTyrLChgObgyJwYtRiPEjXvZStsPSuNvitMUg4b87u-VjIRLjHjad9nXemzfTFc-MBJ5wC58edAZ3SuQoylQz-oR3RvtyAhhihMwxzCdn3etuxHq9K_r1--fh8MFPoADXHgxa1bIt5soCLgATNkQQDtiYqAJ2bCg0dLa9lbBXobRLbiZ4OWCagHFWGC_naBZKAsYfhEMNDmyszJ5wz0K87V2yioltaGVAg5a_cTdl1reUrRDsGYRcBjW1MpARGylrlc4D7lZKBGvBXkQsjuO_cr8QP3ubSxJkj8zsdToR6_Y7x2Ye1NYaVna0M6MdmqKgvVJokVep34XLZqQevoKyg7L3ik4nuNfKKVGo79Ci0IQazp42_rNHQdba6Kj2Rumn0m_70jQj-NCYh380p25MqnnES8BSKEnDu6oup9zNoISUJ7SUMBW-cClP9Zqg1SC_LHXGE4ceIo7G5zOeTEVhKfK1nM3k73ZBKmdwFJ5H_UrWPzszEMk?type=png\)]\(https://mermaid.live/edit#pako:eNqVU01PAjEQ_StNTyrLChgObgyJwYtRiPEjXvZStsPSuNvitMUg4b87u-VjIRLjHjad9nXemzfTFc-MBJ5wC58edAZ3SuQoylQz-oR3RvtyAhhihMwxzCdn3etuxHq9K_r1--fh8MFPoADXHgxa1bIt5soCLgATNkQQDtiYqAJ2bCg0dLa9lbBXobRLbiZ4OWCagHFWGC_naBZKAsYfhEMNDmyszJ5wz0K87V2yioltaGVAg5a_cTdl1reUrRDsGYRcBjW1MpARGylrlc4D7lZKBGvBXkQsjuO_cr8QP3ubSxJkj8zsdToR6_Y7x2Ye1NYaVna0M6MdmqKgvVJokVep34XLZqQevoKyg7L3ik4nuNfKKVGo79Ci0IQazp42_rNHQdba6Kj2Rumn0m_70jQj-NCYh380p25MqnnES8BSKEnDu6oup9zNoISUJ7SUMBW-cClP9Zqg1SC_LHXGE4ceIo7G5zOeTEVhKfK1nM3k73ZBKmdwFJ5H_UrWPzszEMk)

Reviewer 2: @elmiko For posterity, could we put this image in as part of the PR?

If I remember correctly, mermaid.js is also available on both k8s.io and k8s.dev, too.

Author: absolutely

This new initialization process adds some latency to the node readiness. Previously, the kubelet was able to initialize the node at the same time it created the node. Since the logic has moved to the cloud-controller-manager, this can cause a [chicken and egg problem](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#chicken-and-egg) during the cluster bootstrapping for those Kubernetes architectures that do not deploy the controller manager as the other components of the control plane, commonly as static pods, standalone binaries or daemonsets/deployments with tolerations to the taints and using `hostNetwork` (more on this below)

## Examples of the dependency problem

As noted above, it is possible during bootstrapping for the cloud-controller-manager to be unschedulable and as such the cluster will not initialize properly. The following are a few concrete examples of how this problem can be expressed and the root causes for why they might occur.

> These examples assume you are running your cloud-controller-manager using a Kubernetes resource (e.g. `Deployment`, `DaemonSet`, or similar) to control its lifecycle. Because these methods rely on Kubernetes to schedule the cloud-controller-manager, care must be taken to ensure it will schedule properly.

Reviewer: Per the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/): no backticks here.

One of the conditions that can lead to a `Node` resource having this taint is when the container network has not yet been initialized on that node. As the cloud-controller-manager is responsible for adding the IP addresses to a `Node` resource, and the IP addresses are needed by the container network controllers to properly configure the container network, it is possible in some circumstances for a node to become stuck as not ready and uninitialized permanently.

This situation occurs for a similar reason as the first example, although in this case, the `node.kubernetes.io/not-ready` taint is used with the no execute effect and thus will cause the cloud-controller-manager not to run on the node with the taint. If the cloud-controller-manager is not able to execute, then it will not initialize the node. It will cascade into the container network controllers not being able to run properly, and the node will end up carrying both the `node.cloudprovider.kubernetes.io/uninitialized` and `node.kubernetes.io/not-ready` taints, leaving the cluster in an unhealthy state.

> ## Recommendations

Reviewer: Here, this is the authors' opinions. I suggest:
```suggestion
## Our recommendations
```

There is no one “correct way” to run a cloud-controller-manager. The details will depend on the specific needs of the cluster administrators and users. When planning your clusters and the lifecycle of the cloud-controller-managers please consider the following guidance:

For cloud-controller-managers running in the same cluster, they are managing.

> 1. Enable host networking. In most cases, a cloud controller manager will need to communicate with an API service endpoint associated with the infrastructure. Setting “hostNetwork” to true will ensure that the cloud controller is using the host networking instead of the container network and, as such, will have the same network access as the host operating system. It will also remove the dependency on the networking plugin. This will ensure that the cloud controller has access to the infrastructure endpoint (always check your networking configuration against your infrastructure provider’s instructions).  

Reviewer: How I'd write it:
```suggestion
1. Use host network mode, rather than the pod network: in most cases, a cloud controller manager
   will need to communicate with an API service endpoint associated with your infrastructure.
   Setting `hostNetwork` to true ensures that the cloud controller manager is using the node's network interface
   instead of the container network and, as such, will have the same network access as the host operating system.
   Using host network mode also avoids a dependency on the cluster's network plugin(s), and that choice helps to ensure
   that the cloud controller manager can access to infrastructure-relevant API endpoints.  
   Always check your networking configuration against your infrastructure provider’s instructions.
```

3. Target the controller manager containers to the control plane. There might exist other controllers which need to run outside the control plane (for example, Azure’s node manager controller). Still, the controller managers themselves should be deployed to the control plane. Use a node selector or affinity stanza to direct the scheduling of cloud controllers to the control plane to ensure that they are running in a protected space. Cloud controllers are vital to adding and removing nodes to a cluster as they form a link between Kubernetes and the physical infrastructure. Running them on the control plane will help to ensure that they run with a similar priority as other core cluster controllers and that they have some separation from non-privileged user workloads.  
   1. It is worth noting that an anti-affinity stanza to prevent cloud controllers from running on the same host is also very useful to ensure that a single node failure will not degrade the cloud controller performance.  

> 4. Ensure that the tolerations allow operation. Use tolerations on the manifest for the cloud controller container to ensure that it will schedule to the correct nodes and that it can run in situations where a node is initializing. This means that cloud controllers should tolerate the `node.cloudprovider.kubernetes.io/uninitialized` taint, and it should also tolerate any taints associated with the control plane (for example, `node-role.kubernetes.io/master` or `node-role.kubernetes.io/control-plane`). It can also be useful to tolerate the `node.kubernetes.io/not-ready` taint to ensure that the cloud controller can run even when the node is not yet available for health monitoring.

Reviewer: - I'd really like to list `node-role.kubernetes.io/control-plane` ahead of the legacy role `node-role.kubernetes.io/master`.
- Officially, those are labels and not taints. You might be thinking of `node-role.kubernetes.io/control-plane:NoSchedule` and `node-role.kubernetes.io/master:NoSchedule`.

Author: good catch, thank you

Author: actually, for the not-ready, i was thinking about the taint. the one listed here https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions

For cloud-controller-managers that will not be running on the cluster they manage (for example, in a hosted control plane on a separate cluster), then the rules are much more constrained by the dependencies of the environment of the cluster running the cloud-controller-manager. The advice for running on a self-managed cluster may not be appropriate as the types of conflicts and network constraints will be different. Please consult the architecture and requirements of your topology for these scenarios.

### Example

This is an example of a Kubernetes Deployment highlighting the guidance shown above. It is important to note that this is for demonstration purposes only, for production uses please consult your cloud provider’s documentation.

```
apiVersion: apps/v1
kind: Deployment

> metadata:

Reviewer: Consider setting the [`kubernetes.io/description`](https://kubernetes.io/docs/reference/labels-annotations-taints/#description) annotation, per good practice.

>   labels:

Reviewer: Please use the [recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels) here unless there are clear grounds not to.

    k8s-app: cloud-controller-manager
  name: cloud-controller-manager

>   namespace: kube-cloud-controller-manager

Reviewer: This is an unregistered system namespace. Why wouldn't you use `kube-system` here?

Author: a common pattern we see with cloud controller users is putting the ccm in a different namespace. for this example though, i think it makes perfect sense to use `kube-system`. good suggestion

spec:
  replicas: 2
  selector:
    matchLabels:
      k8s-app: cloud-controller-manager
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        k8s-app: cloud-controller-manager
    spec:

>       containers: [] # TODO

Reviewer: You need to fix this before we can publish the article.

>       hostNetwork: true

Reviewer: ```suggestion
      hostNetwork: true # these Pods are part of the control plane
```

      nodeSelector:

>         node-role.kubernetes.io/master: ""

Reviewer: Don't use a deprecated node role in an article we publish to our official blog.

Author: sorry, missed this one. updated.

      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - topologyKey: "kubernetes.io/hostname"
            labelSelector:
              matchLabels:
                k8s-app: cloud-controller-manager
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
        operator: Exists
      - effect: NoExecute
        key: node.kubernetes.io/unreachable
        operator: Exists
        tolerationSeconds: 120
      - effect: NoExecute
        key: node.kubernetes.io/not-ready
        operator: Exists
        tolerationSeconds: 120
      - effect: NoSchedule
        key: node.cloudprovider.kubernetes.io/uninitialized
        operator: Exists
      - effect: NoSchedule
        key: node.kubernetes.io/not-ready
        operator: Exists
```

> 

Reviewer: Would you recommend using cluster-proportional autoscaling for CCM? Maybe you recommend _not_ using automatic scaling here?

Author: no, i definitely recommend running a specific number of CCM replicas. they do not normally benefit from scaling beyond having a duplicate ready for high-availability and redundant protection. i will add a note at the bottom.

\[0\] [https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/\#running-cloud-controller-manager](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)  

> \[1\] [https://kubernetes.io/docs/reference/labels-annotations-taints/\#node-kubernetes-io-not-ready](https://kubernetes.io/docs/reference/labels-annotations-taints/#node-kubernetes-io-not-ready)

Reviewer: We prefer inline-style links; that makes work easier for localization teams.

---
layout: blog
title: "The Cloud Controller Manager Chicken and Egg Problem"
date: 2025-02-14
slug: cloud-controller-manager-chicken-egg-problem
author: >
  Antonio Ojea,
  Michael McCune

> draft: true

Reviewer: ```suggestion
```
We can actually publish this one.

---

Kubernetes 1.31
[completed the largest migration in Kubernetes history][migration-blog], removing the in-tree
cloud provider.  While the component migration is now done, this leaves some additional
complexity for users and installer projects (for example, kOps or Cluster API) .  We will go
over those additional steps and failure points and make recommendations for cluster owners.
This migration was complex and some logic had to be extracted from the core components,
building four new subsystems.

1. **Cloud controller manager** ([KEP-2392][kep2392])
2. **API server network proxy** ([KEP-1281][kep1281])
3. **kubelet credential provider plugins** ([KEP-2133][kep2133])
4. **Storage migration to use [CSI][csi]** ([KEP-625][kep625])

The [cloud controller manager is part of the control plane][ccm]. It is a critical component
that replaces some functionality that existed previously in the kube-controller-manager and the
kubelet.

> ![Components of Kubernetes](https://kubernetes.io/images/docs/components-of-kubernetes.svg)

Reviewer: nit:
- should use a `figure` shortcode
- should use a site-relative link

We can fix these post-publication.

One of the most critical functionalities of the cloud controller manager is the node controller,
which is responsible for the initialization of the nodes.

As you can see in the following diagram, when the **kubelet** starts, it registers the `Node`
object with the apiserver, Tainting the node so it can be processed first by the
cloud-controller-manager. The initial `Node` is missing the cloud-provider specific information,
like the Node Addresses and the Labels with the cloud provider specific information like the

> Node, Region and Instance type information.

Reviewer: nit: no backticks around Node

```mermaid
sequenceDiagram
    autonumber
    rect rgb(191, 223, 255)
    Kubelet->>+Kube-apiserver: Create Node
    Note over Kubelet: Taint:<br/> node.cloudprovider.kubernetes.io
    Kube-apiserver->>-Kubelet: Node  Created
    end
    Note over Kube-apiserver: Node is Not Ready<br/> Tainted, Missing Node Addresses*, ...
    Note over Kube-apiserver: Send Updates
    rect rgb(200, 150, 255)
    Kube-apiserver->>+Cloud-controller-manager: Watch: New Node Created
    Note over Cloud-controller-manager: Initialize Node:<br/>Cloud Provider Labels, Node Addresses, ...
    Cloud-controller-manager->>-Kube-apiserver: Update Node
    end
    Note over Kube-apiserver: Node is Ready

> ```

Reviewer: For maintainability: ideally, capture this to SVG and turn it into a static SVG + `figure` shortcode.

Author: i'm not familiar with the `figure` shortcode, do you have a docs link about it?

Reviewer: See https://github.com/kubernetes/website/pull/49612#issuecomment-2659827569 and https://kubernetes.io/docs/contribute/style/diagram-guide/

schedule taint is not removed, then critical workloads, such as the container network interface
controllers, will not be able to schedule, and the cluster will be left in an unhealthy state.

### Example: Cloud controller manager not scheduling due to not-ready taint

The next example would be possible in situations where the container network interface (CNI) is
waiting for IP address information from the cloud-controller-manager (CCM), and the CCM has not
tolerated the taint which would be removed by the CNI.

The [Kubernetes documentation describes][kubedocs1] the `node.kubernetes.io/not-ready` taint as follows:

> "The Node controller detects whether a Node is ready by monitoring its health and adds or removes this taint accordingly."

One of the conditions that can lead to a `Node` resource having this taint is when the container
network has not yet been initialized on that node. As the cloud-controller-manager is responsible
for adding the IP addresses to a `Node` resource, and the IP addresses are needed by the container
network controllers to properly configure the container network, it is possible in some

> circumstances for a node to become stuck as not ready and uninitialized permanently.

Reviewer: nit: no backticks around Node

  taints associated with the control plane (for example, `node-role.kubernetes.io/control-plane`
  or `node-role.kubernetes.io/master`). It can also be useful to tolerate the
  `node.kubernetes.io/not-ready` taint to ensure that the cloud controller can run even when the
  node is not yet available for health monitoring.

For cloud-controller-managers that will not be running on the cluster they manage (for example,
in a hosted control plane on a separate cluster), then the rules are much more constrained by the
dependencies of the environment of the cluster running the cloud-controller-manager. The advice
for running on a self-managed cluster may not be appropriate as the types of conflicts and network
constraints will be different. Please consult the architecture and requirements of your topology
for these scenarios.

### Example

This is an example of a Kubernetes Deployment highlighting the guidance shown above. It is
important to note that this is for demonstration purposes only, for production uses please
consult your cloud provider’s documentation.

```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: cloud-controller-manager

>     kubernetes.io/description: "Cloud controller manager for my infrastructure"

Reviewer: This is an annotation and won't work as a label. We can fix this post-merge.
